test/parallel/test-worker-locks.js

```diff
@@ -0,0 +1,198 @@
+      firstRequestDone = true;
+    }));
+    await secondRequest;
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+  }
+
+  {
+    // Acquiring a shared lock will not make other shared lock requests for the
+    // same name wait.
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+    await locks.request('lock-name', {
+      mode: 'shared'
+    }, mustCall(async (lock) => {
+      await locks.request('lock-name', {
+        mode: 'shared'
+      }, mustCall(async (lock) => {
+        const snapshot = await locks.query();
+        assert.strictEqual(snapshot.pending.length, 0);
+        assert.strictEqual(snapshot.held.length, 2);
+      }));
+    }));
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+  }
+
+  {
+    // Acquiring a shared lock will make exclusive lock requests for the same
+    // name wait.
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+    let firstRequestDone = false;
+    let secondRequest;
+    let secondRequestDone = false;
+    await locks.request('lock-name', {
+      mode: 'shared'
+    }, mustCall(async (lock) => {
+      secondRequest = locks.request('lock-name', mustCall(async (lock) => {
+        assert(firstRequestDone);
+        secondRequestDone = true;
+        const snapshot = await locks.query();
+        assert.strictEqual(snapshot.pending.length, 0);
+        assert.strictEqual(snapshot.held.length, 1);
+      }));
+      const snapshot = await locks.query();
+      assert.strictEqual(snapshot.pending.length, 1);
+      assert.strictEqual(snapshot.held.length, 1);
+      assert.strictEqual(secondRequestDone, false);
+      await tick();
+      assert.strictEqual(secondRequestDone, false);
+      firstRequestDone = true;
+    }));
+    await secondRequest;
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+  }
+
+  {
+    // Lock requests can be aborted.
+    const ee = new EventEmitter();
+    const locked = locks.request('lock-name', mustCall(async (lock) => {
+      await once(ee, 'stop');
+    }));
+    const controller = new AbortController();
+    const cantlock = locks.request('lock-name', {
+      signal: controller.signal
+    }, mustNotCall());
+    await tick();
+    assert.strictEqual((await locks.query()).held.length, 1);
+    assert.strictEqual((await locks.query()).pending.length, 1);
+    controller.abort();
+    await assert.rejects(cantlock, { name: 'AbortError' });
+    ee.emit('stop');
+    await locked;
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
+  }
+
+  {
+    // Aborting already-acquired locks has no effect.
+    const controller = new AbortController();
+    await locks.request('lock-name', {
+      signal: controller.signal
+    }, mustCall(async (lock) => {
+      controller.abort();
+    }));
+    assert.deepStrictEqual(await locks.query(), { held: [], pending: [] });
```

Comment:
Should test here that the `AbortController` does not hold on to the abort listener in this case

Reply:
If you’re referring to the locks implementation removing the `'abort'` listener, then, done. Otherwise I’m not sure what you mean, the `AbortController` doesn’t hold on to anything here.